lib/srv/db/postgres/redshift-activate-user.sql

```diff
@@ -0,0 +1,40 @@
+create or replace procedure teleport_activate_user(username varchar, roles text)
+language plpgsql
+as $$
+declare
+    roles_length integer;
+    cur_roles_length integer;
+begin
+    roles_length := JSON_ARRAY_LENGTH(roles);
+
+    -- If the user already exists and was provisioned by Teleport, reactivate
+    -- it, otherwise provision a new one.
+    if exists (select user_id from svv_user_grants where user_name = username and admin_option = false and role_name = 'teleport-auto-user') then
+        -- If the user has active connections, make sure the provided roles
+        -- match what the user currently has.
+        if exists (select user_name from stv_sessions where user_name = concat('IAM:', username)) then
+          select into cur_roles_length count(role_name) from svv_user_grants where user_name = username and admin_option=false and role_name != 'teleport-auto-user';
+          if roles_length != cur_roles_length then
+            raise exception 'User has active connections and roles have changed';
+          end if;
```

Comment:
We only count the number of roles here; I think this is important that we do an exact comparison.

Reply:
it's checked in the next loop

Reply:
Sorry, I missed that. I guess this is an optimization, then?

Reply:
the next loop checks if all new roles are currently assigned. But if there are more current roles, the length check will fail. didn't find a good way to aggregate and compare in one shot so have to do this length check + loop combo.